scripts/builtin/decisionTree.dml

```diff
@@ -0,0 +1,449 @@
+#                                       - R[1,]: Row Vector which indicates if feature vector is scalar or categorical. 1 indicates
+#                                     a scalar feature vector, other positive Integers indicate the number of categories
+#                                     If R is not provided by default all variables are assumed to be scale
+# bins          Int      20           Number of equiheight bins per scale feature to choose thresholds
+# depth         Int      25           Maximum depth of the learned tree
+# M             String   ---          Location to write matrix M containing the learned tree
+# ---------------------------------------------------------------------------------------------
+# OUTPUT:
+# Matrix M where each column corresponds to a node in the learned tree and each row contains the following information:
+#    M[1,j]: id of node j (in a complete binary tree)
+#    M[2,j]: Offset (no. of columns) to left child of j if j is an internal node, otherwise 0
+#    M[3,j]: Feature index of the feature (scale feature id if the feature is scale or categorical feature id if the feature is categorical)
+#            that node j looks at if j is an internal node, otherwise 0
+#    M[4,j]: Type of the feature that node j looks at if j is an internal node: holds the same information as R input vector
+#    M[5,j]: If j is an internal node: 1 if the feature chosen for j is scale, otherwise the size of the subset of values
+#            stored in rows 6,7,... if j is categorical
+#            If j is a leaf node: number of misclassified samples reaching at node j
+#    M[6:,j]: If j is an internal node: Threshold the example's feature value is compared to is stored at M[6,j] if the feature chosen for j is scale,
+#             otherwise if the feature chosen for j is categorical rows 6,7,... depict the value subset chosen for j
+#             If j is a leaf node 1 if j is impure and the number of samples at j > threshold, otherwise 0
+# -------------------------------------------------------------------------------------------
+
+m_decisionTree = function(
+    Matrix[Double] X,
+    Matrix[Double] Y,
+    Matrix[Double] R,
+    int bins = 10,
+    int depth = 20
+)  return (Matrix[Double] M) {
+    #calcPossibleThresholdsCategory(1)
+    node_queue = matrix(1, rows=1, cols=1)             # Add first Node
+    impurity_queue = matrix(1, rows=1, cols=1)
+    use_cols_queue = matrix(1, rows=ncol(X), cols=1)   # Add fist bool Vector with all cols <=> (use all cols)
+    use_rows_queue = matrix(1, rows=nrow(X), cols=1)   # Add fist bool Vector with all rows <=> (use all rows)
+    queue_length = 1
+    M = matrix(0, rows = 0, cols = 0)
+    while (queue_length > 0) {
+       print("-------------------------------------------------------------------------------------------------------")
+        [node_queue, node] = dataQueuePop(node_queue)
```

Comment:
indentation is not consistant. i suggest 2 spaces indent for dml scrips (since we use this in most of the files.)